pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -78,6 +81,55 @@ func AddToManager(mgr manager.Manager,
 	return nil
 }
 
+type simpleImageStream struct {
+	cluster, namespace, name string
+}
+
+type shardedLock struct {
+	mapLock *sync.Mutex
+	locks   map[simpleImageStream]*sync.Mutex
+}
+
+func (s *shardedLock) getLock(key simpleImageStream) *sync.Mutex {
+	s.mapLock.Lock()
+	defer s.mapLock.Unlock()
+	if _, exists := s.locks[key]; !exists {
+		s.locks[key] = &sync.Mutex{}
+	}
+	return s.locks[key]
+}
+
+// cleanup deletes all locks by acquiring first
+// the mapLock and then each individual lock before
+// deleting it. The individual lock must be acquired
+// because otherwise it may be held, we delete it from
+// the map, it gets recreated and acquired and two
+// routines report in parallel for the same job.
+// Note that while this function is running, no new
+// presubmit reporting can happen, as we hold the mapLock.
+func (s *shardedLock) cleanup() {
+	s.mapLock.Lock()
+	defer s.mapLock.Unlock()
+
+	for key, lock := range s.locks {
+		lock.Lock()
+		delete(s.locks, key)
+		lock.Unlock()
+	}
+}
+
+// runCleanup asynchronously runs the cleanup once per hour.
+func (s *shardedLock) runCleanup() {
+	go func() {
+		for range time.Tick(time.Hour) {
+			logrus.Debug("Starting to clean up presubmit locks")
```

Comment:
Please use a logger that has the controller name and update the logs here and the godoc comment above to not refer to "presubmits" anymore, that will be wildly confusing later on :)

Reply:
Done.